// railway/pg/index.js
'use strict';

var pg = require('pg');
var fs = require('fs');
var path = require('path');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var pg__default = /*#__PURE__*/_interopDefaultLegacy(pg);
var fs__default = /*#__PURE__*/_interopDefaultLegacy(fs);
var path__default = /*#__PURE__*/_interopDefaultLegacy(path);

/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */

var __assign = function() {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

var railwayEnvBuildPath = path__default['default'].resolve(process.cwd(), "./.railway/env.json");
var nodeModulesEnv = null;
try {
    nodeModulesEnv = require("./env.json");
}
catch (e) {
    // do nothing
}
var getEnvFromFile = function () {
    if (nodeModulesEnv != null) {
        return nodeModulesEnv;
    }
    if (fs__default['default'].existsSync(railwayEnvBuildPath)) {
        return JSON.parse(fs__default['default'].readFileSync(railwayEnvBuildPath, "utf8"));
    }
    return {};
};
var builtConfig = getEnvFromFile();
process.env = __assign(__assign({}, builtConfig), process.env);
var client = new pg__default['default'].Pool();
var query = function (statement, cb) {
    client.query(statement, cb);
};
var connect = function () {
    client.connect();
};
var close = function () {
    client.end();
};
var postgres = {
    query: query,
    close: close,
    connect: connect,
    client: client,
    config: {
        host: process.env.PGHOST,
        port: process.env.PGPORT,
        user: process.env.PGUSER,
        password: process.env.PGPASSWORD,
        database: process.env.PGDATABASE,
        url: process.env.DATABASE_URL,
    },
};

module.exports = postgres;
